linkerd/retry/src/replay.rs

```diff
@@ -0,0 +1,766 @@
+
+/// Wraps an HTTP body type and lazily buffers data as it is read from the inner
+/// body.
+///
+/// When this body is dropped, if a clone exists, any buffered data is shared
+/// with its cloned. The first clone to be polled will take ownership over the
+/// data until it is dropped. When *that* clone is dropped, the buffered data
+/// --- including any new data read from the body by the clone, if the body has
+/// not yet completed --- will be shared with any remaining clones.
+///
+/// The buffered data can then be used to retry the request if the original
+/// request fails.
+#[derive(Debug)]
+pub struct ReplayBody<B> {
+    /// Buffered state owned by this body if it is actively being polled. If
+    /// this body has been polled and no other body owned the state, this will
+    /// be `Some`.
+    state: Option<BodyState<B>>,
+
+    /// Copy of the state shared across all clones. When the active clone is
+    /// dropped, it moves its state back into the shared state to be taken by the
+    /// next clone to be polled.
+    shared: Arc<Mutex<Option<BodyState<B>>>>,
+
+    /// Should this clone replay the buffered body from the shared state before
+    /// polling the initial body?
+    replay_body: bool,
+
+    /// Should this clone replay trailers from the shared state?
+    replay_trailers: bool,
+}
+
+/// Data returned by `ReplayBody`'s `http_body::Body` implementation is either
+/// `Bytes` returned by the initial body, or a list of all `Bytes` chunks
+/// returned by the initial body (when replaying it).
+#[derive(Debug)]
+pub enum Data {
+    Initial(Bytes),
+    Replay(BufList),
+}
+
+/// Body data composed of multiple `Bytes` chunks.
+#[derive(Clone, Debug, Default)]
+pub struct BufList {
+    bufs: VecDeque<Bytes>,
+}
+
+/// A `Service`/`Proxy` that wraps an HTTP request's `Body` type in a `ReplayBody`,
+/// allowing it to be cloned.
+///
+// TODO(eliza): it would be nice if this could just be `MapRequest`, but Tower
+// won't let us implement `Proxy` for it (because it doesn't expose access to
+// the closure). Maybe we can change that upstream eventually...
+#[derive(Debug, Default, Clone)]
+pub struct WrapBody<S> {
+    inner: S,
+}
+
+#[derive(Debug)]
+struct BodyState<B> {
+    buf: BufList,
+    trailers: Option<HeaderMap>,
+    rest: Option<B>,
+    is_completed: bool,
+}
+
+// === impl ReplayBody ===
+
+impl<B: Body> ReplayBody<B> {
+    /// Wraps an initial `Body` in a `ReplayBody`.
+    pub fn new(body: B) -> Self {
+        Self {
+            state: Some(BodyState {
+                buf: Default::default(),
+                trailers: None,
+                rest: Some(body),
+                is_completed: false,
+            }),
+            shared: Arc::new(Mutex::new(None)),
+            // The initial `ReplayBody` has nothing to replay
+            replay_body: false,
+            replay_trailers: false,
+        }
+    }
+
+    fn state<'a>(
```

Comment:
perhaps `acquire_state` would be a little more descriptive? or `get_or_insert_from`? my concern is that `state` doesn't necessarily make it obvious that this takes a value out of the shared lock.